### ISSUE TITLE
Atomic ordering in lammps-NN

### ISSUE THREAD

**Issue by qzhu2017** [qzhu2017/PyXtal_FF#37]
@yanxon 
I just created another example for [binary system](https://github.com/qzhu2017/PyXtal_FF/tree/master/examples/SiO2-LAMMPS)

It closely followed what we did for Si. However, the `validate.py` does not work.
```
qzhu@cms SiO2-LAMMPS (master) $ python validate.py 

ASE calculator with pyxtal_ff force field

Energy:  -73.011 eV/atom
Forces (eV/A)
[[ -0.3004   0.5202   0.0000]
 [ -0.3004  -0.5202   0.0000]
 [  0.6007  -0.0000   0.0000]
 [ -0.3956  -0.1307  -0.7050]
 [  0.3110  -0.2772  -0.7050]
 [  0.0846   0.4080  -0.7050]
 [  0.0846  -0.4080   0.7050]
 [ -0.3956   0.1307   0.7050]
 [  0.3110   0.2772   0.7050]]
Stresses (GPa)
[ -0.0607  -0.0607  -0.0683  -0.0000  -0.0000   0.0000]

LAMMPS calculator with pyxtal_ff force field

Energy:  -92.435 eV/atom
Forces (eV/A)
[[ -0.3835   0.6642   0.0000]
 [ -0.3835  -0.6642   0.0000]
 [  0.7670  -0.0000   0.0000]
 [  0.0593   0.9024  -1.3199]
 [ -0.8112  -0.3998  -1.3199]
 [  0.7519  -0.5026  -1.3199]
 [  0.7519   0.5026   1.3199]
 [  0.0593  -0.9024   1.3199]
 [ -0.8112   0.3998   1.3199]]
Stresses (GPa)
[  0.0992   0.0992   0.1870   0.0000  -0.0000   0.0000]
```

In the [debug](https://github.com/qzhu2017/PyXtal_FF/tree/master/examples/SiO2-LAMMPS/debug) folder,

I added the script to check the descriptor output from pyxtal_ff and lammps. It seems that they are consistent. Therefore, I suspect there exist some issues in mliap_nn to deal with the prediction of multicomponent systems.

I think this is related to the way how we arrange the atomic orders. Please take a look and fix this issue.

**Comment by qzhu2017** [comment]
@pedroantoniosantosf @yanxon  
Can you check why the mliap_NN returns different results for total energy?

**Comment by qzhu2017** [comment]
I am debugging it now. For this quartz structure. Pyxtal_FF output the following energy
```
Total Eng-O:  -64.35585169546353 (-10.726 per atom)
Total Eng-Si:  -8.654828344193852 (-2.884 per atom)
Total Energy:  -73.011 eV
```
While lammps outputs
```
O: -10.726 eV/atom
Si: -0.35987 eV/atom
```

**Comment by pedroantoniosantosf** [comment]
> I am debugging it now. For this quartz structure. Pyxtal_FF output the following energy
> 
> ```
> Total Eng-O:  -64.35585169546353 (-10.726 per atom)
> Total Eng-Si:  -8.654828344193852 (-2.884 per atom)
> Total Energy:  -73.011 eV
> ```
> 
> While lammps outputs
> 
> ```
> O: -10.726 eV/atom
> Si: -0.35987 eV/atom
> ```
> 
> It seems that the `coeffelem[0]` and `coeffelem[1]` contains the same weight parameters.
> https://github.com/lammps/lammps/blob/56e3962d9c4c7560e06c16b5c2475063e1abb47b/src/MLIAP/mliap_model_nn.cpp#L224
> 
> Therefore, the energy prediction of `Si`is actually based on the weight parameter of `O`.
> 
> @pedroantoniosantosf Can you double check the following func works properly for multicomponent?
> 
> https://github.com/lammps/lammps/blob/56e3962d9c4c7560e06c16b5c2475063e1abb47b/src/MLIAP/mliap_model_nn.cpp#L67-L209

It seems that the weight parameters are the same for Si and O. Let me check with other file.

**Comment by pedroantoniosantosf** [comment]
With the current [binary system](https://github.com/qzhu2017/PyXtal_FF/tree/master/examples/SiO2-LAMMPS/SiO2-snap)
I got this result:

```
Total energy (eV) = 
Total energy (eV/atom) = -10.2706063442014 
```

**Comment by yanxon** [comment]
> It seems that the `coeffelem[0]` and `coeffelem[1]` contains the same weight parameters.
> https://github.com/lammps/lammps/blob/56e3962d9c4c7560e06c16b5c2475063e1abb47b/src/MLIAP/mliap_model_nn.cpp#L224
> 
> Therefore, the energy prediction of `Si`is actually based on the weight parameter of `O`.> 

I don't think this is the case. I changed the `lt_quartz.data` to all `O`:
```
lt_quartz.data (written by ASE) 

9 	 atoms 
2  atom types
0.0      5.0277820499999999  xlo xhi
0.0       4.354186979991403  ylo yhi
0.0      5.5189180000000002  zlo zhi
    -2.5138910249999986                       0                       0  xy xz yz

Masses

1 10
2 20

Atoms 

     1   1     -1.1998550473222493      2.0782099036800967      1.8396393149369399
     2   1      1.3140359776777506      2.2759770763113063      5.5189179816036065
     3   1         2.3997100946445                       0      3.6792786482702735
     4   1      -1.066450392298574      3.2469302935405291        2.96550046336906
     5   1     0.23519210262592588      1.8071487016146919      1.1258611300357273
     6   1     0.83125828967265147      3.6542949648275855      4.8051397967023934
     7   1      3.3451493146726508     0.69989201516381805      2.5534175366309402
     8   1      1.4474406327014251      1.1072566864508737      4.3930568699642736
     9   1       2.749083127625926      2.5470382783767112     0.71377820329760655
```

Results:
```
Total energy (eV)      = -109.640764609725 
Total energy (eV/atom) = -12.1823071788583 
```

Whereas, I change all atoms to `Si`:
```
Total energy (eV)      = -99.2703439628473 
Total energy (eV/atom) = -11.0300382180941 
```

**Comment by qzhu2017** [comment]
https://github.com/qzhu2017/PyXtal_FF/blob/master/examples/SiO2-LAMMPS/SiO2-snap/NN_weights.txt
```
Line 24
# Coefficients
  -0.0564130565110   -0.1882714347708   -0.2206491358092   -0.2685284141367    0.1821186806033 
  -0.1513666440473    0.0735912428690    0.2666858673159   -0.1772220461340   -0.0200233500257 
  -0.0138071094916   -0.0136877976481   -0.1166629665395   -0.2059563372188   -0.5293614040911 
  -0.3250920726825   -0.1790875400560    0.0088982478387   -0.3023740519026   -0.2104025833620 
  -0.4894199842268   -0.1327970724406   -0.0337041103308   -0.1634876134813   -0.1645665543345 
  -0.0547292732608   -0.2483194577897   -0.1534921843190    0.1909776782569   -0.0829732502269 
  -0.1594464470767    0.4032005190755    0.0955973991757   -0.3236864584898   -0.0896706553063 
```

```
line 421
# Coefficients
  -0.0564130565110   -0.1882714347708   -0.2206491358092   -0.2685284141367    0.1821186806033 
  -0.1513666440473    0.0735912428690    0.2666858673159   -0.1772220461340   -0.0200233500257 
  -0.0138071094916   -0.0136877976481   -0.1166629665395   -0.2059563372188   -0.5293614040911 
  -0.3250920726825   -0.1790875400560    0.0088982478387   -0.3023740519026   -0.2104025833620 
  -0.4894199842268   -0.1327970724406   -0.0337041103308   -0.1634876134813   -0.1645665543345 
  -0.0547292732608   -0.2483194577897   -0.1534921843190    0.1909776782569   -0.0829732502269 
  -0.1594464470767    0.4032005190755    0.0955973991757   -0.3236864584898   -0.0896706553063 
  -0.0337271484141   -0.0251271025873   -0.0869340782420   -0.0114470504023   -0.1616005503215 
```

@pedroantoniosantosf @yanxon they are the same.

@yanxon Can you double check if you did something wrong when dumping the weight parameters?

**Comment by yanxon** [comment]
@qzhu2017 @pedroantoniosantosf 

Yes. I messed up. Please give me half an hour to fix this.

**Comment by yanxon** [comment]
@qzhu2017 

Can you please try it again?

**Comment by qzhu2017** [comment]
@yanxon 
@pedroantoniosantosf 
I just made a couple of changes based on @yanxon 's fix. Now it works. Nothing wrong on mliap-nn

```
qzhu@cms SiO2-LAMMPS (master) $ python validate.py 

ASE calculator with pyxtal_ff force field

Energy:  -73.625 eV
Forces (eV/A)
[[ -0.3741   0.6480   0.0000]
 [ -0.3741  -0.6480   0.0000]
 [  0.7482  -0.0000   0.0000]
 [  0.0891   0.0423  -0.9700]
 [ -0.0812   0.0560  -0.9700]
 [ -0.0079  -0.0984  -0.9700]
 [ -0.0079   0.0984   0.9700]
 [  0.0891  -0.0423   0.9700]
 [ -0.0812  -0.0560   0.9700]]
Stresses (GPa)
[  0.0222   0.0222   0.0216  -0.0000   0.0000   0.0000]

LAMMPS calculator with pyxtal_ff force field

Energy:  -73.625 eV
Forces (eV/A)
[[ -0.3741   0.6480   0.0000]
 [ -0.3741  -0.6480   0.0000]
 [  0.7482  -0.0000   0.0000]
 [  0.0891   0.0423  -0.9700]
 [ -0.0812   0.0560  -0.9700]
 [ -0.0079  -0.0984  -0.9700]
 [ -0.0079   0.0984   0.9700]
 [  0.0891  -0.0423   0.9700]
 [ -0.0812  -0.0560   0.9700]]
Stresses (GPa)
[  0.0222   0.0222   0.0216   0.0000  -0.0000   0.0000]
```

**Comment by qzhu2017** [comment]
Now fully resolved after 96b83ac
